Pre-Releases for Breaking Releases (RC1) (#2832)
ethereumjs/ethereumjs-monorepo: packages/block/CHANGELOG.md

(modification: no type change headlines) and this project adheres to
[Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## 5.0.0-rc.1 - 2023-07-11

### Default Shanghai HF / Merge -> Paris Renaming / Cancun Hardfork

The Shanghai hardfork is now the default HF in `@ethereumjs/common` and therefore for all libraries who use a Common-based HF setting internally (e.g. Tx, Block or EVM), see PR [#2655](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2655).

Also the Merge HF has been renamed to Paris (`Hardfork.Paris`) being the correct HF name on the execution side, see [#2652](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2652). To set the HF to Paris in Common you can do:

```typescript
import { Chain, Common, Hardfork } from '@ethereumjs/common'
const common = new Common({ chain: Chain.Mainnet, hardfork: Hardfork.Paris })
```

> And third on hardforks 🙂: while not all Cancun EIPs are finalized yet, Cancun is now an officially selectable hardfork in our libraries (see PR [#2659](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2659)) and can be activated with:

Author: Yes, that makes a lot of sense, added.

```typescript
import { Chain, Common, Hardfork } from '@ethereumjs/common'
const common = new Common({ chain: Chain.Mainnet, hardfork: Hardfork.Cancun })
```

### New API to set Hardforks

> Our APIs to (re-)set a a hardfork within a library had grown old over all changes on how this is done over the years. 😂

Reviewer: ```suggestion
We have notably simplified our APIs to (re-)set a a hardfork within a library, which had grown old over the years. 😂
```

```typescript
common.setHardforkBy({ blockNumber: 5000000n }) // Setting a mainnet common to a Block from `Byzantium` (and so: to `Byzantium` HF)
common.setHardforkBy({ timestamp: 1681340000n }) // Setting a mainnet common to a post-Shanghai timestamp
common.setHardforkBy({ blockNumber, timestamp }) // Setting a common with to a not pre-known HF using both block number and timestamp
```

There is a third option `td` which is Merge specific and should normally not be used except for a very rare set of dynamic Merge-HF scenarios.

For the `Block` library we also updated the old concurrent `hardforkByBlockNumber` and `hardforkByTTD` options to a unified and simplified `setHardfork` option for the constructor, see PR [#2800](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2800).

### API Validation Methods Clean-Up

We have cleaned up and unified the validation methods in the `Block` library, see PR [#2792](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2792).

The `Block.validateTransactions()` method, previously overloaded with different return types depending on the input, has been split up into:

```typescript
Block.transactionsAreValid(): boolean

> Block.getTransactionsValidationErrors(): string[] // If you are interested in the errors, can also be used for validation by checking return array length

Author: I have the weakest opinion here for anything 😋, I will just skip any recommendation and remove the comment.

```

Other renamings:

```typescript
Block.validateTransactionsTrie(): Promise<boolean> // old
Block.transactionsTrieIsValid(): Promise<boolean> // new

Block.validateUnclesHash(): boolean // old
Block.uncleHashIsValid(): boolean // new

Block.validateWithdrawalsTrie(): Promise<boolean> // old
Block.withdrawalsTrieIsValid(): Promise<boolean> // new
```

### EIP-4844 Support (Status: Review, 4844-devnet-7, July 2023)

> While there might be last-round final tweaks [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844) is closing in on its final format with a lot of spec changes during the last 2-3 months still happening.

Author: Have added locally (and will push later).

> While there might be last-round final tweaks [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844) is closing in on its final format with a lot of spec changes during the last 2-3 months still happening.

Author: Have rephrased with:

```shell
While there might be last-round final tweaks [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844) is closing in on its final format. A lot of spec changes happened during the last 2-3 months and these are included in this release round. So the released version should be relatively close to a future production ready version.
```

Hope this works better. 🙂

> While there might be last-round final tweaks [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844) is closing in on its final format with a lot of spec changes during the last 2-3 months still happening.

Reviewer: I find this much clearer! Thanks :)

> While there might be last-round final tweaks [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844) is closing in on its final format with a lot of spec changes during the last 2-3 months still happening.

Author: Yeah, I "wanted a bit too much" with the previous one-sentence version. 😂

This release supports EIP-4844 along this snapshot [b9a5a11](https://github.com/ethereum/EIPs/commit/b9a5a117ab7e1dc18f937841d00598b527c306e7)from the EIP repository with the EIP being in `Review` status and features/changes included which made it into [4844-devnet-7](https://github.com/ethpandaops/4844-testnet).

#### KZG Initialization -> @ethereumjs/util

The global initialization method for the KZG setup has been moved to a dedicated [kzg.ts](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/util/src/kzg.ts) module in `@ethereumjs/util` for easy reuse across the libraries, see PR [#2567](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2567).

The `initKZG()` method can be used as follows:

```typescript
// Make the kzg library available globally
import * as kzg from 'c-kzg'
import { initKZG } from '@ethereumjs/util'

// Initialize the trusted setup
initKZG(kzg, 'path/to/my/trusted_setup.txt')
```

For further information on this see the respective section in `@ethereumjs-util` [README](https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/util).

#### New Block Header field: dataGasUsed

> For the Block library the most significant change is that there is now a new header field `dataGasUsed` if `EIP-4844` is activated (via `Common`), see PR [#2750](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2750).

Author: This was already added in the original 4844 implementation - see https://github.com/ethereumjs/ethereumjs-monorepo/pull/2349. So I am not giving a full 4844 summary here, this is just about the stuff that changed from the last released 4844 version.

Additionally there are the following three `dataGasUsed` related new helper methods:

```typescript
BlockHeader.getDataGasPrice(): bigint
BlockHeader.calcDataFee(numBlobs: number): bigint

> BlockHeader.calcNextExcessDataGas(): bigint

Author: Ah, thanks, yes. 👍 I have now changed the whole header to:

```shell
#### New Block Header field: dataGasUsed / new excessDataGas Helpers
```

and also updated the sentence above this method summary block.

```

#### Other Library Changes

- Update eip4844 blocks/txs to decoupled blobs spec, PR [#2567](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2567)

### Block: fromExecutionPayload() / fromBeaconPayloadJson()

Two new handy constructors have been added to the `Block` class to bring the consensus and execution parts of Ethereum closer together:

> `Block.fromBeaconPayloadJson()` allows to initialize an Ethereum L1 block with a payload received from the beacon chain via an RPC call. 🤩 The new constructor can be used as follows:

Author: Ah, `EL` is likely the better term here? Will update.

```typescript
const block = await Block.fromBeaconPayloadJson(payload, { common })
```

If you already have the execution payload embedded within a beacon block you can alternatively directly use the `Block.fromExecutionPayload()` constructor.

### Hybrid CJS/ESM Build

We now provide both a CommonJS and an ESM build for all our libraries. 🥳 This transition was a huge undertaking and should make the usage of our libraries in the browser a lot more straight-forward, see PR [#2685](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2685), [#2783](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2783), [#2786](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2786), [#2764](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2764), [#2804](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2804) and [#2809](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2809) (and others). We rewrote the whole set of imports and exports within the libraries, updated or completely removed a lot of dependencies along the way and removed the usage of all native Node.js primitives (like `https` or `util`).

> There are now two different build directories in our `dist` folder, being `dist/cjs` for the CommonJS and `dist/esm` for the `ESM` build. That means that direct imports (which you generally should try to avoid, rather open an issue on your import needs), needs an update within your code (do a `dist` or the like code search).

Author: Will apply locally, other occasions as well, won't give separate comments.

Both builds have respective separate entrypoints in the distributed `package.json` file.

A CommonJS import of our libraries can then be done like this:

```typescript
const { Chain, Common } = require('@ethereumjs/common')
const common = new Common({ chain: Chain.Mainnet })
```

And this is how an ESM import looks like:

```typescript
import { Chain, Common } from '@ethereumjs/common'
const common = new Common({ chain: Chain.Mainnet })
```

Using ESM will give you additional advantages over CJS beyond browser usage like static code analysis / Tree Shaking which CJS can not provide.

Side note: along this transition we also rewrote our whole test suite (yes!!!) to now work with [Vitest](https://vitest.dev/) instead of `Tape`.

### Buffer -> Uint8Array

> With this releases we remove all Node.js specific `Buffer` usages from our libraries and replace these with [Uint8Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array) representations, which are available both in Node.js and the browser (`Buffer` is a subclass of `Uint8Array`). While this is a big step towards interoperability and browser compatibility of our libraries, this is also one of the most invasive operations we have ever done, see the huge changeset from PR [#2566](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2566) and [#2607](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2607). 😋

Reviewer: ```suggestion
With this release we remove all Node.js specific `Buffer` usages from our libraries and replace these with [Uint8Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array) representations, which are available both in Node.js and the browser (`Buffer` is a subclass of `Uint8Array`). While this is a big step towards interoperability and browser compatibility of our libraries, this is also one of the most invasive operations we have ever done, see the huge changeset from PR [#2566](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2566) and [#2607](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2607). 😋
```

Author: Small nit: I would want to have this in "plural", so that people are aware that this is a change which is done in all other libraries as well

(side note: you have actually done this change in 4/12 instances. This is only semi-helpful TBH)

#### How to upgrade?

For this library you should check if you use one of the following constructors, methods, constants or types and do a search and update input and/or output values or general usages and add conversion methods if necessary:

```typescript
// header
BlockHeader.fromHeaderData(headerData: HeaderData = {}, opts: BlockOptions = {})
BlockHeader.fromRLPSerializedHeader(serializedHeaderData: Uint8Array, opts: BlockOptions = {})
BlockHeader.fromValuesArray(values: BlockHeaderBytes, opts: BlockOptions = {})
new BlockHeader() // deprecated
BlockHeader.raw(): BlockHeaderBytes
BlockHeader.hash(): Uint8Array
BlockHeader.cliqueExtraVanity(): Uint8Array // clique PoA only
BlockHeader.cliqueExtraSeal(): Uint8Array // clique PoA only
BlockHeader.serialize(): Uint8Array

// block
Block.fromBlockData(blockData: BlockData = {}, opts?: BlockOptions)
Block.fromRLPSerializedBlock(serialized: Uint8Array, opts?: BlockOptions)
Block.fromValuesArray(values: BlockBytes, opts?: BlockOptions)
new Block() // deprecated
Block.raw(): BlockBytes
Block.hash(): Uint8Array
Block.serialize(): Uint8Array
```

> We have added helper methods for "Buffer -> Uint8Array" conversions in the [@ethereumjs/util](https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/util) `bytes` module, see the respective README section for guidance.

Reviewer 2: I don't think this is 100% accurate. There are no buffer to bytes conversion methods in util and longer, or did I miss them somewhere?

Reviewer: Yeah seconding @acolytec3 's point, to my knowledge we have no helper method that accepts Buffer as input. Would perhaps be more accurate to write that:
" We have converted existing Buffer conversion methods to Uint8Array conversion methods [...]  " ?

Author: Ah, yes, I already stumbled upon this upon writing but then I though "ah, hopefully they'll get it" 😂.

Will take the version from Gabriel, yes, that's more accurate.

#### Prefixed Hex Strings as Default

> It is a constant source of errors and mismatches in byte-handling code bases if hex strings are used in a mixed prefixed/unprefixed way.

Author: Have updated locally.

We have therefore decided to go "prefixed" by default, see PR [#2830](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2830) and [#2845](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2845).

The methods `hexToBytes` and `bytesToHex`, also similar methods like `intToHex` now both take in prefixed hex strings and give prefixed strings as an output. The corresponding unprefixed methods are marked as `deprecated` and usage should be avoided.

Please therefore check your code base on updating and ensure that values you are passing to constructors and methods are prefixed with a `0x`.

### Other Changes

- Support for `Node.js 16` has been removed (minimal version: `Node.js 18`), PR [#2859](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2859)
- Migrate `fromEthersProvider()` to more general `fromJsonRpcProvider()` constructor, remove `Ethers` dependency, PR [#2663](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2663)
- Breaking: `Block._common` property has been renamed to `Block.common`, PR [#2857](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2857)

> - While the main `new Block()` constructor should generally not be used (use the static constructor methods instead) just to mention: `withdrawals` and `opts` constructor arguments have been swapped, PR [#2715](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2715)

Reviewer 2: ```suggestion
- While the main `new Block()` constructor should generally not be used (use the static constructor methods instead) just to mention: the order of the `withdrawals` and `opts` constructor arguments has been swapped, PR [#2715](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2715)
```
